File: src/org/wavescale/sourcesync/action/ActionSelectedFilesToRemote.java
package org.wavescale.sourcesync.action;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.wavescale.sourcesync.api.ConnectionConfiguration;
import org.wavescale.sourcesync.api.FileSynchronizer;
import org.wavescale.sourcesync.api.SynchronizationQueue;
import org.wavescale.sourcesync.api.Utils;
import org.wavescale.sourcesync.factory.ConfigConnectionFactory;
import org.wavescale.sourcesync.factory.ConnectionConfig;
import org.wavescale.sourcesync.logger.BalloonLogger;
import org.wavescale.sourcesync.logger.EventDataLogger;

import java.io.File;
import java.nio.file.Path;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Semaphore;

/**
 * ****************************************************************************
 * Copyright (c) 2014-2107 Faur Ioan-Aurel.                                     *
 * All rights reserved. This program and the accompanying materials             *
 * are made available under the terms of the MIT License                        *
 * which accompanies this distribution, and is available at                     *
 * http://opensource.org/licenses/MIT                                           *
 * *
 * For any issues or questions send an email at: fioan89@gmail.com              *
 * *****************************************************************************
 */
public class ActionSelectedFilesToRemote extends AnAction {
    public void actionPerformed(final AnActionEvent e) {
        // first check if there's a connection type associated to this module.
        // If not alert the user and get out
        Project project = PlatformDataKeys.PROJECT.getData(e.getDataContext());

        String projectName = project.getName();
        String associationName = ConnectionConfig.getInstance().getAssociationFor(projectName);
        if (associationName == null) {
            Utils.showNoConnectionSpecifiedError(projectName);
            return;
        }

        // get a list of selected virtual files
        VirtualFile[] virtualFiles = PlatformDataKeys.VIRTUAL_FILE_ARRAY.getData(e.getDataContext());
        if (virtualFiles.length <= 0) {
            StringBuilder builder = new StringBuilder("Project <b>");
            builder.append(e.getProject().getName()).append("</b>! does not have files selected!");
            BalloonLogger.logBalloonInfo(builder.toString(), e.getProject());
            EventDataLogger.logInfo(builder.toString(), e.getProject());
            return;
        }

        // start sync
        final ConnectionConfiguration connectionConfiguration = ConfigConnectionFactory.getInstance().
                getConnectionConfiguration(associationName);
        final Semaphore semaphores = new Semaphore(connectionConfiguration.getSimultaneousJobs());
        final int allowed_sessions = virtualFiles.length <= connectionConfiguration.getSimultaneousJobs() ?
                virtualFiles.length : connectionConfiguration.getSimultaneousJobs();
        final SynchronizationQueue synchronizationQueue = new SynchronizationQueue(e.getProject(), connectionConfiguration, allowed_sessions);
        synchronizationQueue.startCountingTo(virtualFiles.length);
        final BlockingQueue<FileSynchronizer> queue = synchronizationQueue.getSyncQueue();
        for (final VirtualFile virtualFile : virtualFiles) {
            if (virtualFile != null && new File(virtualFile.getPath()).isFile()) {
                if (Utils.canBeUploaded(virtualFile.getName(), connectionConfiguration.getExcludedFiles())) {
                    final Path uploadLocation = Utils.dirsToFileFromProjectRoot(virtualFile, project);
                    ProgressManager.getInstance().run(new Task.Backgroundable(e.getProject(), "Uploading", false) {
                        @Override
                        public void run(@NotNull ProgressIndicator indicator) {
                            FileSynchronizer fileSynchronizer = null;
                            try {
                                semaphores.acquire();
                                fileSynchronizer = queue.take();
                                fileSynchronizer.setIndicator(indicator);
                                if (fileSynchronizer != null) {
                                    fileSynchronizer.connect();
                                    // so final destination will look like this:
                                    // root_home/ + project_relative_path_to_file/
                                }
                                fileSynchronizer.syncFile(virtualFile.getPath(), uploadLocation);
                                queue.put(fileSynchronizer);
                                synchronizationQueue.count();
                            } catch (InterruptedException e1) {
                                e1.printStackTrace();
                            } finally {
                                semaphores.release();
                            }
                        }
                    });

                } else {
                    if (virtualFile != null) {
                        EventDataLogger.logWarning("File <b>" + virtualFile.getName() + "</b> is filtered out!", e.getProject());
                        synchronizationQueue.count();
                    }
                }
            } else {
                if (virtualFile != null) {
                    EventDataLogger.logWarning("File <b>" + virtualFile.getName() + "</b> is a directory!", e.getProject());
                    synchronizationQueue.count();
                }
            }
        }
    }
}

File: src/main/kotlin/org/wavescale/sourcesync/action/ActionSelectedFilesToRemote.kt
package org.wavescale.sourcesync.action

import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.ProjectManager
import com.intellij.project.stateStore
import com.intellij.ui.NewUI
import java.io.File
import org.wavescale.sourcesync.SourceSyncIcons
import org.wavescale.sourcesync.SourcesyncBundle
import org.wavescale.sourcesync.api.Utils
import org.wavescale.sourcesync.configurations.ScpSyncConfiguration
import org.wavescale.sourcesync.configurations.SshSyncConfiguration
import org.wavescale.sourcesync.configurations.SyncConfigurationType
import org.wavescale.sourcesync.notifications.Notifier
import org.wavescale.sourcesync.services.SyncRemoteConfigurationsService
import org.wavescale.sourcesync.synchronizer.SCPFileSynchronizer
import org.wavescale.sourcesync.synchronizer.SFTPFileSynchronizer
import org.wavescale.sourcesync.synchronizer.Synchronizer

class ActionSelectedFilesToRemote : AnAction() {
    private val syncConfigurationsService = ProjectManager.getInstance().openProjects[0].getService(SyncRemoteConfigurationsService::class.java)
    override fun getActionUpdateThread() = ActionUpdateThread.BGT

    override fun actionPerformed(e: AnActionEvent) {
        // first check if there's a connection type associated to this module.
        // If not alert the user and get out
        val project = e.project ?: return

        // get a list of selected virtual files
        val virtualFiles = PlatformDataKeys.VIRTUAL_FILE_ARRAY.getData(e.dataContext)!!
        if (virtualFiles.isEmpty()) {
            Notifier.notifyInfo(
                e.project!!,
                SourcesyncBundle.message("no.files.selected.to.sync")
            )
            return
        }

        val mainConfiguration = syncConfigurationsService.mainConnection()
        if (mainConfiguration == null) {
            Notifier.notifyError(
                e.project!!,
                SourcesyncBundle.message("no.remote.sync.connection.configured.title"),
                SourcesyncBundle.message("no.remote.sync.connection.configured.message")
            )
            return
        }

        val fileSynchronizer: Synchronizer = when (mainConfiguration.protocol) {
            SyncConfigurationType.SCP -> {
                SCPFileSynchronizer(mainConfiguration as ScpSyncConfiguration, project)
            }

            SyncConfigurationType.SFTP -> {
                SFTPFileSynchronizer(mainConfiguration as SshSyncConfiguration, project)
            }
        }
        var directoriesFound = false
        val (files, rest) = virtualFiles.filterNotNull().partition { File(it.path).isFile }
        rest.forEach {
            directoriesFound = true
            logger.info("Skipping upload of ${it.name} because it's a directory")
        }

        if (directoriesFound) {
            Notifier.notifyUpgradeToProDueToFolderUpload(project)
        }

        val (acceptedFiles, excludedFiles) = files.partition { Utils.canBeUploaded(it.name, mainConfiguration.excludedFiles) }
        excludedFiles.forEach {
            logger.info("Skipping upload of ${it.name} because it matches the exclusion file pattern")
        }
        ProgressManager.getInstance().run(object : Task.Backgroundable(e.project, "Uploading", false) {
            override fun run(indicator: ProgressIndicator) {
                try {
                    if (fileSynchronizer.connect()) {
                        fileSynchronizer.syncFiles(acceptedFiles.map { Pair(it.path, Utils.relativeToProjectPath(it, project.stateStore)) }.toSet(), indicator)
                    }
                } finally {
                    fileSynchronizer.disconnect()
                }
            }
        })
    }

    override fun update(e: AnActionEvent) {
        super.update(e)
        if (NewUI.isEnabled()) {
            e.presentation.icon = SourceSyncIcons.ExpUI.SOURCESYNC
        }

        val mainConnectionName = syncConfigurationsService.mainConnectionName()
        if (mainConnectionName != null) {
            e.presentation.apply {
                text = "Sync selected files to $mainConnectionName"
                isEnabled = true
            }
        } else {
            e.presentation.apply {
                text = "Sync selected files to Remote target"
                isEnabled = false
            }
        }
    }

    companion object {
        val logger = logger<ActionSelectedFilesToRemote>()
    }
}